### PR TITLE
Feature/8 implement methods

### PR DESCRIPTION
**Pull request by miroslavpojer**
#8 - Implement method ideas
**Implemented full expected methods for Log 'ResponseAction'.**
> def logError(message: String): Boolean
> def logWarn(message: String): Boolean
> def logInfo(message: String): Boolean
> def logDebug(message: String): Boolean

**Implemented methods for Log 'ResponseAction'.**
- response-time-...

> def assertResponseTimeIsBelow(response: Response, limit: String): Boolean
> def assertResponseTimeIsAbove(response: Response, limit: String): Boolean

- status-code-...

> def assertStatusCodeEquals(response: Response): Boolean
> def assertStatusCodeSuccess(response: Response): Boolean
> def assertStatusCodeIsClientError(response: Response): Boolean
> def assertStatusCodeIsServerError(response: Response): Boolean

- header-...

> def assertHeaderExists(response: Response, headerName: String): Boolean
> def assertHeaderValueEquals(response: Response, headerName: String, expectedValue: String): Boolean

- content-type-...

> def assertContentTypeIsJson(response: Response): Boolean
> def assertContentTypeIsXml(response: Response): Boolean
> def assertContentTypeIsHtml(response: Response): Boolean

- cookie-...

> def assertCookieExists(response: Response, cookieName: String): Boolean
> def assertCookieValueEquals(response: Response, cookieName: String, expectedValue: String): Boolean
> def assertCookieIsSecured(response: Response, cookieName: String): Boolean
> def assertCookieIsNotSecured(response: Response, cookieName: String): Boolean

- body-.
> def assertBodyContainsText(response: Response, text: String): Boolean

**Done several refactoring changes:**
- improved logging messages
- Set to Seq to keep defined actions order.
- small improvements solved when found

### CODE REVIEW

File: testApi/src/main/scala/africa/absa/testing/scapi/model/Method.scala [hunk 1]
```diff
@@ -27,9 +27,9 @@ import africa.absa.testing.scapi.json.{Action, Header, Requestable, ResponseActi
  * @param responseActions The set of responseAction objects for the method.
  */
 case class Method(name: String,
-                  headers: Set[Header],
-                  actions: Set[Action],
-                  responseActions: Set[ResponseAction]) extends Requestable {
+                  headers: Seq[Header],
```
[the remark this answers is not in the document]
Comment (miroslavpojer) [reply testApi/src/main/scala/africa/absa/testing/scapi/model/Method.scala:30]: The presented change is here to keep items in order defined in source json file. When the Runtime Cache feature introduced then the steps could use cache values saved in previous steps. The Seq type is here to keep this order.

File: testApi/src/main/scala/africa/absa/testing/scapi/rest/request/sender/ScAPIRequestSender.scala [hunk 2]
```diff
@@ -18,11 +18,35 @@ package africa.absa.testing.scapi.rest.request.sender
...
 
+import java.net.HttpCookie
+
 /**
  * ScAPIRequestSender is an implementation of the RequestSender interface.
  * It provides the capability to send different types of HTTP requests including GET, POST, PUT, and DELETE.
  */
 object ScAPIRequestSender extends RequestSender {
+
+  private def sendRequest(requestFunc: => requests.Response, url: String, headers: Map[String, String], verifySslCerts: Boolean, data: String, params: Map[String, String]): Response = {
```
[the remark this answers is not in the document]
Comment (miroslavpojer) [reply testApi/src/main/scala/africa/absa/testing/scapi/rest/request/sender/ScAPIRequestSender.scala]: Fixed in commit 6032a6431c2645db83b934a09e6dd6d90a719282.

File: testApi/src/main/scala/africa/absa/testing/scapi/rest/response/Response.scala [hunk 3]
```diff
@@ -56,14 +57,34 @@ object Response {
    * @return           Boolean indicating whether all response actions passed (true) or any response action failed (false).
    * @throws IllegalArgumentException If an response action group is not supported.
    */
-  def perform(response: Response, responseAction: Set[ResponseAction]): Boolean = {
+  def perform(response: Response, responseAction: Seq[ResponseAction]): Boolean = {
+    def logParameters(response: Response, resolvedResponseAction: ResponseAction, exception: Option[Throwable] = None): Unit = {
+      val filteredParams = resolvedResponseAction.params.filter(_._1 != "method").map { case (k, v) => s"$k->$v" }.mkString(", ")
+      val baseLog = s"\nParameters received: \n\tRequired Response-Action: \n\t\tGroup->'${resolvedResponseAction.group}', \n\t\tMethod->'${resolvedResponseAction.name}', \n\t\tParams->'${filteredParams}', \n\tActual Response: \n\t\t$response"
```
[the remark this answers is not in the document]
Comment (miroslavpojer) [reply testApi/src/main/scala/africa/absa/testing/scapi/rest/response/Response.scala]: Fixed in commit - 6032a6431c2645db83b934a09e6dd6d90a719282.

File: testApi/src/main/scala/africa/absa/testing/scapi/rest/response/LogResponseAction.scala [hunk 4]
```diff
@@ -54,10 +57,13 @@ object LogResponseAction extends ResponsePerformer {
    * @throws IllegalArgumentException if the response action's name is not recognized.
    */
   def performResponseAction(response: Response, responseAction: ResponseAction): Boolean = {
```
[the remark this answers is not in the document]
Comment (miroslavpojer) [reply testApi/src/main/scala/africa/absa/testing/scapi/rest/response/LogResponseAction.scala:59]: The returned boolean value is consumed in logic of SuiteRunner where this information is used for SuiteResults purposes.  This could be replaced be throwing an exception when logic failed.

[the remark this answers is not in the document]
Comment (Zejnilovic) [reply testApi/src/main/scala/africa/absa/testing/scapi/rest/response/LogResponseAction.scala:59]: So my idea would be:
- Remove all the `true` from the small `logSomething` methods. 
- Replace it here with something better, like that `SuiteResult.Success`
- Replace that throw with `SuiteResult.Failure`

How does that sound?

File: testApi/src/main/scala/africa/absa/testing/scapi/rest/response/LogResponseAction.scala [hunk 5]
```diff
@@ -54,10 +57,13 @@ object LogResponseAction extends ResponsePerformer {
...
+    val message = responseAction.params.getOrElse("message", throw new IllegalArgumentException("Missing 'message' parameter"))
+
     responseAction.name match {
-      case INFO =>
-        val message = responseAction.params("message")
-        logInfo(message)
+      case ERROR => logError(message)
+      case WARN => logWarn(message)
+      case INFO => logInfo(message)
+      case DEBUG => logDebug(message)
```
Comment (dk1844) [inline testApi/src/main/scala/africa/absa/testing/scapi/rest/response/LogResponseAction.scala:66]: Again, these should be probably enum/sealed case objects. You would get match exhaustion checking from the compiler with the sealed case object, e.g.

File: testApi/src/main/scala/africa/absa/testing/scapi/rest/response/Response.scala [hunk 6]
```diff
@@ -17,8 +17,9 @@
 package africa.absa.testing.scapi.rest.response
 
 import africa.absa.testing.scapi.json.ResponseAction
+import africa.absa.testing.scapi.logging.Logger
 
-case class Response(statusCode: Int, body: String, headers: Map[String, Seq[String]])
+case class Response(statusCode: Int, body: String, url: String, statusMessage: String, headers: Map[String, Seq[String]], cookies: Map[String, (String, Boolean)], duration: Long)
```
Comment (dk1844) [inline testApi/src/main/scala/africa/absa/testing/scapi/rest/response/Response.scala:22]: Imho the cookie type `Map[String, (String, Boolean)]` does not communicate the intention well; then you have to do abysmal calls like:
`response.cookies(cookieName)._1` where you would rather want to do `response.cookies(cookieName).value`

Therefore: 
```suggestion
case class CookieValue(value: String, secured: Boolean)
case class Response(statusCode: Int, body: String, url: String, statusMessage: String, headers: Map[String, Seq[String]], cookies: Map[String, CookieValue], duration: Long)
```

**Review by dk1844 — CHANGES_REQUESTED** [review]
Not part of this PR, but the Logger with var `level` getting set around is very weird.

Some of the intercept[xxx] call do not check the error message, wouldn't it be worth being present everywhere? Or are there some exceptions?

File: testApi/src/main/scala/africa/absa/testing/scapi/rest/response/LogResponseAction.scala (hunk 4 above)
[the remark this answers is not in the document]
Comment (miroslavpojer) [reply testApi/src/main/scala/africa/absa/testing/scapi/rest/response/LogResponseAction.scala]: Fixed in commit c203ab6e0a2670215c2b1f00296202274cfbd634.

File: testApi/src/main/scala/africa/absa/testing/scapi/rest/response/LogResponseAction.scala [hunk 7]
```diff
@@ -66,6 +72,26 @@ object LogResponseAction extends ResponsePerformer {
...
+  }
+
+  /**
+   * This method logs a message at the WARN level.
+   *
+   * @param message The message to be logged.
+   */
+  def logWarn(message: String): Boolean = {
+    Logger.warn(message)
+    true
```
[the remark this answers is not in the document]
Comment (miroslavpojer) [reply testApi/src/main/scala/africa/absa/testing/scapi/rest/response/LogResponseAction.scala]: Fixed in commit Fixed in commit c203ab6e0a2670215c2b1f00296202274cfbd634..

File: testApi/src/main/scala/africa/absa/testing/scapi/rest/response/Response.scala [hunk 8]
```diff
@@ -56,14 +57,42 @@ object Response {
...
-      resolvedResponseAction.group match {
-        case GROUP_ASSERT => AssertionResponseAction.performResponseAction(response, assertion)
-        case GROUP_EXTRACT_JSON => ExtractJsonResponseAction.performResponseAction(response, assertion)
-        case GROUP_LOG => LogResponseAction.performResponseAction(response, assertion)
-        case _ => throw new IllegalArgumentException(s"Unsupported assertion group: ${assertion.group}")
+      Logger.debug(s"Response-${resolvedResponseAction.group}: '${resolvedResponseAction.name}' - Started.")
+
+      try {
+        val res = resolvedResponseAction.group match {
+          case GROUP_ASSERT => AssertionResponseAction.performResponseAction(response, assertion)
```
[the remark this answers is not in the document]
Comment (miroslavpojer) [reply testApi/src/main/scala/africa/absa/testing/scapi/rest/response/Response.scala]: Fixed in commit c203ab6e0a2670215c2b1f00296202274cfbd634.

File: testApi/src/main/scala/africa/absa/testing/scapi/utils/validation/ContentValidator.scala [hunk 9]
```diff
@@ -37,6 +37,19 @@ object ContentValidator {
...
+  /**
+   * Validates that a string can be parsed to a long. Throws an exception if the string cannot be parsed.
+   *
+   * @param input The string to be validated.
+   * @throws ContentValidationFailed if the input string cannot be parsed to a long.
+   */
+  def validateLongString(input: String, param: String): Unit = {
+    Try(input.toLong) match {
+      case Success(_) => // Do nothing
+      case Failure(e) => throw ContentValidationFailed(input, s"Received value of '$param' cannot be parsed to a long: ${e.getMessage}")
```
[the remark this answers is not in the document]
Comment (miroslavpojer) [reply testApi/src/main/scala/africa/absa/testing/scapi/utils/validation/ContentValidator.scala]: I would like to keep it here.

File: testApi/src/main/scala/africa/absa/testing/scapi/rest/response/LogResponseAction.scala (hunk 5 above)
Comment (miroslavpojer) [reply testApi/src/main/scala/africa/absa/testing/scapi/rest/response/LogResponseAction.scala:65]: Fixed in commit 2e251aeeb8d90762628f8d6456e2a9c08e9d882d.

**Comment by miroslavpojer** [comment]
> Not part of this PR, but the Logger with var `level` getting set around is very weird.
> 
> Some of the intercept[xxx] call do not check the error message, wouldn't it be worth being present everywhere? Or are there some exceptions?

All intercepts reviewed and now all of them should check message too.

File: testApi/src/main/scala/africa/absa/testing/scapi/rest/response/Response.scala (hunk 6 above)
Comment (miroslavpojer) [reply testApi/src/main/scala/africa/absa/testing/scapi/rest/response/Response.scala]: Fixed in commit 11293dcc6b1762816cfcd832e4fa189b42882327 and one before.

**Comment by miroslavpojer** [comment]
I have finished all review notes. See commits in review note answers.
I have done big amount of refactoring changes. Please @dk1844 and @Zejnilovic to do fully new review of current state.

File: testApi/src/main/scala/africa/absa/testing/scapi/rest/response/LogResponseAction.scala (hunk 7 above)
[the remark this answers is not in the document]
Comment (dk1844) [reply testApi/src/main/scala/africa/absa/testing/scapi/rest/response/LogResponseAction.scala]: I think the `log*` methods should just return `Unit`. `Try[X]` in their signature would indicate that they could fail at what they do which is simply not true.

File: testApi/src/main/scala/africa/absa/testing/scapi/rest/response/action/ExtractJsonResponseAction.scala [hunk 10]
```diff
@@ -77,17 +81,17 @@ object ExtractJsonResponseAction extends ResponsePerformer {
    * @param listIndex            The index in the JSON array from which to extract the string.
    * @param jsonKey              The key in the JSON object from which to extract the string.
    * @param runtimeCacheLevel    The expiration level to use when storing the extracted string in the runtime cache.
-   * @return Boolean indicating whether the string extraction and caching operation was successful.
+   * @return A Try[Unit] indicating whether the string extraction and caching operation was successful or not.
    */
-  def stringFromList(response: Response, cacheKey: String, listIndex: Int, jsonKey: String, runtimeCacheLevel: String): Boolean = {
+  private def stringFromList(response: Response, cacheKey: String, listIndex: Int, jsonKey: String, runtimeCacheLevel: String): Try[Unit] = {
     try {
```
Comment (dk1844) [inline testApi/src/main/scala/africa/absa/testing/scapi/rest/response/action/ExtractJsonResponseAction.scala:87]: why `try` to return `Success()`, why not just `Try {code-that-may-throw-otherwise we end up with Success)`?

File: testApi/src/main/scala/africa/absa/testing/scapi/rest/response/action/LogResponseAction.scala [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+
+  /**
+   * Logs a message at the DEBUG level.
+   *
+   * @param message The message to be logged.
+   * @return A Try[Unit] indicating the success of the logging operation.
+   */
+  private def logDebug(message: String): Try[Unit] = {
+    Logger.debug(message)
+    Success(())
```
Comment (dk1844) [inline testApi/src/main/scala/africa/absa/testing/scapi/rest/response/action/LogResponseAction.scala:117]: as outline, pls no Try[Unit] here.

File: testApi/src/main/scala/africa/absa/testing/scapi/rest/response/action/types/LogResponseActionType.scala [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+  val WARN: LogResponseActionType.Value = Value("warn")
+  val INFO: LogResponseActionType.Value = Value("info")
+  val DEBUG: LogResponseActionType.Value = Value("debug")
+
+  private val stringToValueMap = values.map(v => v.toString -> v).toMap
+
+  def fromString(s: String): Option[LogResponseActionType] = stringToValueMap.get(s)
+
+  // Implicit conversion from AssertResponseActionType to String
+  implicit def enumValueToString(value: LogResponseActionType): String = value.toString
```
Comment (dk1844) [inline testApi/src/main/scala/africa/absa/testing/scapi/rest/response/action/types/LogResponseActionType.scala:34]:
```suggestion
  // Implicit conversion from LogResponseActionTypeto String
  implicit def enumValueToString(value: LogResponseActionType): String = value.toString
```

The comment does not match the implementation.
Also, the `enumValueToString` has a name that does not clarify the usecase fully (rather e.g. `logResponseActionType2String`), but `implicit def`s are frowned upon anyway, so maybe it should not be present at all? Or if you must have it, please move it to the test package. e.g. directly to the test `ResponseAssertionsTest` where it seems to be actually used.

File: testApi/src/main/scala/africa/absa/testing/scapi/rest/response/action/types/AssertResponseActionType.scala [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+
+  // body-...
+  val BODY_CONTAINS_TEXT: AssertResponseActionType.Value = Value("body-contains-text")
+
+  private val stringToValueMap = values.map(v => v.toString -> v).toMap
+
+  def fromString(s: String): Option[AssertResponseActionType] = stringToValueMap.get(s)
+
+  // Implicit conversion from AssertResponseActionType to String
+  implicit def enumValueToString(value: AssertResponseActionType): String = value.toString
```
Comment (dk1844) [inline testApi/src/main/scala/africa/absa/testing/scapi/rest/response/action/types/AssertResponseActionType.scala:57]: dtto

File: testApi/src/main/scala/africa/absa/testing/scapi/rest/response/action/types/ExtractJsonResponseActionType.scala [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+  type ExtractJsonResponseActionType = Value
+
+  val STRING_FROM_LIST: ExtractJsonResponseActionType.Value = Value("string-from-list")
+
+  private val stringToValueMap = values.map(v => v.toString -> v).toMap
+
+  def fromString(s: String): Option[ExtractJsonResponseActionType] = stringToValueMap.get(s)
+
+  // Implicit conversion from AssertResponseActionType to String
+  implicit def enumValueToString(value: ExtractJsonResponseActionType): String = value.toString
```
Comment (dk1844) [inline testApi/src/main/scala/africa/absa/testing/scapi/rest/response/action/types/ExtractJsonResponseActionType.scala:31]: dtto

File: testApi/src/test/scala/africa/absa/testing/scapi/json/SuiteFactoryTest.scala [hunk 15]
```diff
@@ -129,43 +130,43 @@ class SuiteFactoryTest extends FunSuite {
...
 
-    val filteredSuiteBundles: Set[SuiteBundle] = SuiteFactory.filterOnlyOrAll(suitesBundles)
+    val filteredSuiteBundles: Set[Suite] = SuiteFactory.filterOnlyOrAll(suitesBundles)
 
     assertEquals(filteredSuiteBundles.size, 1)
 
     val filteredSuite = filteredSuiteBundles.head.suite
-    assertEquals(filteredSuite.endpoint, "endpoint1")
-    assertEquals(filteredSuite.tests.size, 1)
+    assertEquals(clue("name1"), clue(filteredSuite.name))
```
Comment (dk1844) [inline testApi/src/test/scala/africa/absa/testing/scapi/json/SuiteFactoryTest.scala:150]: What is the point of calling `clue` here (and everywhere else), sorry, I don't get it

File: testApi/src/test/scala/africa/absa/testing/scapi/reporter/StdOutReporterTest.scala [hunk 16]
```diff
@@ -81,11 +83,11 @@ class StdOutReporterTest extends FunSuite {
     val output = baos.toString
 
     // Assertions
-    assertEquals(clue(output.contains("Simple Text Report")), true)
-    assertEquals(clue(output.contains("Number of tests run: 0")), true)
-    assertEquals(clue(output.contains("Number of successful tests: 0")), true)
-    assertEquals(clue(output.contains("Number of failed tests: 0")), true)
-    assertEquals(clue(output.contains("End Report")), true)
+    assertEquals(true, clue(output.contains("Simple Text Report")))
```
Comment (dk1844) [inline testApi/src/test/scala/africa/absa/testing/scapi/reporter/StdOutReporterTest.scala:86]: Why the change in params order?

AFAIK the signature of `assertEquals` expects the actual value to be first and the expected to be second.

```scala
  def assertEquals[A, B](
      obtained: A,
      expected: B,
      clue: => Any = "values are not the same"
  )
```

File: testApi/src/test/scala/africa/absa/testing/scapi/rest/response/ResponseExtractTest.scala [hunk 17]
```diff
@@ -47,124 +66,93 @@ class ResponseExtractTest extends FunSuite {
...
   /*
-    validateStringFromList
+    performResponseAction
    */
 
-  // positive test "stringFromList - correct parameters" - tested during "validateContent - STRING_FROM_LIST"
+  test("performAssertion - STRING_FROM_LIST") {
+    val result: Try[Unit] = ExtractJsonResponseAction.performResponseAction(responseWithID, assertionStringFromList)
+    assert(result.isSuccess)
+    assertEquals("382be85a-1f00-4c15-b607-cbda03ccxxx2", RuntimeCache.get("question_id").get)
```
Comment (dk1844) [inline testApi/src/test/scala/africa/absa/testing/scapi/rest/response/ResponseExtractTest.scala:129]: dtto, swapped values (actual, expected)?

**Review by dk1844 — COMMENTED** [review]
Looks better, thanks for the fixes from last time.

File: testApi/src/main/scala/africa/absa/testing/scapi/rest/response/LogResponseAction.scala (hunk 7 above)
[the remark this answers is not in the document]
Comment (miroslavpojer) [reply testApi/src/main/scala/africa/absa/testing/scapi/rest/response/LogResponseAction.scala]: Fixed in commit e2fec8f3d140ec059a477083233993167a414e5e.

File: testApi/src/main/scala/africa/absa/testing/scapi/suite/runner/SuiteRunner.scala [hunk 18]
```diff
@@ -193,47 +185,61 @@ object SuiteRunner {
...
    * @param e             The exception to handle.
-   * @param suiteEndpoint Suite's endpoint.
+   * @param suiteName     Suite's name.
    * @param name          The name of the suite or test.
    * @param testStartTime The starting time of the suite or test.
    * @param resultType    The type of the suite or test ("Before", "Test", or "After").
    * @return SuiteResults after the exception handling.
    */
-  private def handleException(e: Throwable, suiteEndpoint: String, name: String, testStartTime: Long, resultType: String, categories: Option[String] = None): SuiteResults = {
+  private def handleException(e: Throwable, suiteName: String, name: String, testStartTime: Long, resultType: String, categories: Option[String] = None): SuiteResult = {
```
[the remark this answers is not in the document]
Comment (miroslavpojer) [reply testApi/src/main/scala/africa/absa/testing/scapi/suite/runner/SuiteRunner.scala]: Fixed in commit 470efbf47cb176cacc35e7a6844b09088c352756.

File: testApi/src/main/scala/africa/absa/testing/scapi/rest/response/action/LogResponseAction.scala (hunk 11 above)
Comment (miroslavpojer) [reply testApi/src/main/scala/africa/absa/testing/scapi/rest/response/action/LogResponseAction.scala]: Fixed in commit e2fec8f3d140ec059a477083233993167a414e5e.

File: testApi/src/main/scala/africa/absa/testing/scapi/rest/response/action/ExtractJsonResponseAction.scala (hunk 10 above)
Comment (miroslavpojer) [reply testApi/src/main/scala/africa/absa/testing/scapi/rest/response/action/ExtractJsonResponseAction.scala]: Fixed in commit a15a4201defe36c53ed748e66e4bad34aa6372ac.

File: testApi/src/main/scala/africa/absa/testing/scapi/rest/response/action/types/LogResponseActionType.scala (hunk 12 above)
Comment (miroslavpojer) [reply testApi/src/main/scala/africa/absa/testing/scapi/rest/response/action/types/LogResponseActionType.scala]: Fixed in commit 5ec5a73137e30cedef0c0cab5593948d1f27d33c.

File: testApi/src/main/scala/africa/absa/testing/scapi/rest/response/action/types/AssertResponseActionType.scala (hunk 13 above)
Comment (miroslavpojer) [reply testApi/src/main/scala/africa/absa/testing/scapi/rest/response/action/types/AssertResponseActionType.scala]: Fixed in commit 5ec5a73137e30cedef0c0cab5593948d1f27d33c.

File: testApi/src/main/scala/africa/absa/testing/scapi/rest/response/action/types/ExtractJsonResponseActionType.scala (hunk 14 above)
Comment (miroslavpojer) [reply testApi/src/main/scala/africa/absa/testing/scapi/rest/response/action/types/ExtractJsonResponseActionType.scala]: Fixed in commit 5ec5a73137e30cedef0c0cab5593948d1f27d33c.

File: testApi/src/test/scala/africa/absa/testing/scapi/json/SuiteFactoryTest.scala (hunk 15 above)
Comment (miroslavpojer) [reply testApi/src/test/scala/africa/absa/testing/scapi/json/SuiteFactoryTest.scala:150]: I found an error made from several test changes. Correct usage of Clue is in simple boolean assert method call.

- All location with clue used were refactored.
- Why to use it? Here is comparison of output "when error happen":

**without clue**
image

**with clue** 
image

You can see that clue method provide additional information. It can be useful for  complex objects.

File: testApi/src/test/scala/africa/absa/testing/scapi/reporter/StdOutReporterTest.scala (hunk 16 above)
Comment (miroslavpojer) [reply testApi/src/test/scala/africa/absa/testing/scapi/reporter/StdOutReporterTest.scala:86]: Wrong order is a bug. Fixed in commit **TBD**.

File: testApi/src/test/scala/africa/absa/testing/scapi/rest/response/ResponseExtractTest.scala (hunk 17 above)
Comment (miroslavpojer) [reply testApi/src/test/scala/africa/absa/testing/scapi/rest/response/ResponseExtractTest.scala:133]: Wrong order is a bug. Fixed in commit **TBD**.

File: testApi/src/main/scala/africa/absa/testing/scapi/suite/runner/SuiteRunner.scala (hunk 18 above)
[the remark this answers is not in the document]
Comment (dk1844) [reply testApi/src/main/scala/africa/absa/testing/scapi/suite/runner/SuiteRunner.scala]: Its nice, I just think that `UPPER_SNAKE_CASE_THAT_IS_USED_FOR_RESULT_TYPE_ENUM_VALUES` conforms to Scala conventions https://docs.scala-lang.org/style/naming-conventions.html. They should probably be written in `TheUsualScalaCamelCase` instead.

File: testApi/src/test/scala/africa/absa/testing/scapi/json/SuiteFactoryTest.scala (hunk 15 above)
Comment (dk1844) [reply testApi/src/test/scala/africa/absa/testing/scapi/json/SuiteFactoryTest.scala]: Thanks, got it.

File: testApi/src/test/scala/africa/absa/testing/scapi/reporter/StdOutReporterTest.scala (hunk 16 above)
Comment (dk1844) [reply testApi/src/test/scala/africa/absa/testing/scapi/reporter/StdOutReporterTest.scala]: The signature of methods in _what_ used framework uses opposite order of parameters? MUnit?

I am looking https://scalameta.org/munit/docs/tests.html and I can see things like
`assertEquals(actual, expected) // true`

Also, looking into the code of `munit.Assertions` I see nowhere for this common order to be changed.

Could you elaborate on your reasoning that back the view of the order of obtained-expected to differ, please?

Comment (miroslavpojer) [reply testApi/src/test/scala/africa/absa/testing/scapi/reporter/StdOutReporterTest.scala]: - My experience with first "expected", then "actual" is nicely defined here - https://sourceforge.net/p/junit/mailman/message/3338997/.
- After latest refactoring no more `assertEquals` used.
- How to deal with it future?
image
  - On picture you can see them munit expected 1st actual, 2nd expected.
  - We can following it when assert methods will be used.

Comment (dk1844) [reply testApi/src/test/scala/africa/absa/testing/scapi/reporter/StdOutReporterTest.scala]: After a lengthy discussion offline, it seems that multiple opinions may have merit.

File: testApi/src/test/scala/africa/absa/testing/scapi/rest/response/ResponseExtractTest.scala (hunk 17 above)
Comment (miroslavpojer) [reply testApi/src/test/scala/africa/absa/testing/scapi/rest/response/ResponseExtractTest.scala:133]: After deeper analysis I have decided to follow the framework order to avoid wrong outputs.

File: testApi/src/test/scala/africa/absa/testing/scapi/reporter/StdOutReporterTest.scala (hunk 16 above)
Comment (miroslavpojer) [reply testApi/src/test/scala/africa/absa/testing/scapi/reporter/StdOutReporterTest.scala]: After deeper analysis I have decided to follow the framework order to avoid wrong outputs.

File: testApi/src/main/scala/africa/absa/testing/scapi/suite/runner/SuiteRunner.scala (hunk 18 above)
[the remark this answers is not in the document]
Comment (miroslavpojer) [reply testApi/src/main/scala/africa/absa/testing/scapi/suite/runner/SuiteRunner.scala]: Fixed in commit 4e0d029d469d328ef6092c7a0947d3e6deab8015.

**Review by dk1844 — APPROVED** [review]
LGTM
